# THE EFFECT OF A POSITIVE SHOCK TO TECHNOLOGY IN HIGH- AND LOW-INCOME COUNTRIES

# Introduction

The aim of this project is to analyse the effect of a positive schock to the technological stage for different high- and low-income countries. More specifically, we want to estimate the long-run effects of an increase in technology by 20% from 2022 onwards. This high increase occurs due to the new AI technologies, which has been changing the technological scene all over the world.

**On our data:**

For the analysis, we have collected data from 1960 for gross domestic product $(Q_t)$, physical capital $(K_t)$, techological stage $(B_t)$ and labor force $(L_t)$. The selected countries are Denmark, The United States, India and China. All data is simulated up until year 2100 based on an extended Solow-Model, where all parameter values has been calculated for each country. This means that the dataset is fictional and based on a theoretical model, which does not include outliers or fluctuations. All data from 1960 has been collected from The World Bank and Penn World Table 9.0. More on data can be found at the end of the project in the bibliography on data.

**Methodology:**

We define output by a simple Cobb-Douglas production function:

$Q_{i,t} = B_{i,t}^{\alpha} (K_{i,t}L_{i,t})^{(1-\alpha)}$

Where $i$ denotes the specific country and $t$ the year. When calculating the evolution in $\alpha$-values, we solve for $\alpha$:

$\alpha_{i,t} = \frac{ln(Q_{i,t})-ln(K_t L_t)}{ln(B_t)-ln(K_t L_t)}$

# Initializing the project

**Imports:**

First, we are specifying a CSV file, printing its contents, loading it into a DataFrame, and initializing a Solowclass instance. Then, we import all necessary installations for the project:

In [14]:
# Import necessary libraries
import pandas as pd  # Import pandas for data manipulation
import numpy as np  # Import numpy for numerical computing
import matplotlib.pyplot as plt  # Import matplotlib for plotting
import ipywidgets as widgets  # Import ipywidgets for interactive widgets
from ipywidgets import interact  # Import interact from ipywidgets
from IPython.display import display  # Import display function from IPython.display for displaying output

# Import custom module dataproject for additional functionalities
# Ensure dataproject is accessible in the Python path or working directory
import dataproject  

# Import the AlphaCalculator class from the dataproject module
from dataproject import AlphaCalculator 

# Specifying the filename of the CSV file containing the data
filename = "DataProject_Data.csv"

# Read the CSV file into a pandas DataFrame
# Ensure the delimiter matches the file format
DataProject_Data = pd.read_csv(filename, delimiter=";")

# Create an instance of AlphaCalculator with the provided data
alpha_calculator = AlphaCalculator(DataProject_Data)

# Create an instance of the Solowclass from the dataproject module with the specified filename
solow_instance = dataproject.Solowclass(filename)


**Read and clean data:**

Since the dataset is a CSV file, some necessary cleaning and restructuring of the dataset is necessary, in order to use the data in the project. We need to convert columns into strings, before replacing commas with dots, to be able to convert the data into floats. 

In [ ]:
# Remove trailing whitespaces from column names
DataProject_Data.columns = DataProject_Data.columns.str.strip()

# Convert specific columns to strings before replacing commas
string_columns = ['DK_K_t', 'US_K_t']  # Specify columns to be converted to strings
for column in string_columns:
    DataProject_Data[column] = DataProject_Data[column].astype(str)

# Replace commas with dots in selected columns to prepare for conversion to float
DataProject_Data['DK_K_t'] = DataProject_Data['DK_K_t'].str.replace(',', '.')
DataProject_Data['US_K_t'] = DataProject_Data['US_K_t'].str.replace(',', '.')

# Convert strings to floats for specified columns
columns_to_convert = ['DK_B_t', 'US_B_t', 'IN_B_t', 'CN_B_t',  # Specify columns to be converted to floats
                      'DK_K_t', 'US_K_t', 'IN_K_t', 'CN_K_t',
                      'DK_Q_t', 'US_Q_t', 'IN_Q_t', 'CN_Q_t',
                      'DK_L_t', 'US_L_t', 'IN_L_t', 'CN_L_t']

for column in columns_to_convert:
    # Replace commas with dots and convert the column to float type
    DataProject_Data[column] = DataProject_Data[column].str.replace(',', '.').astype(float)

NameError: name 'DataProject_Data' is not defined

In the next cell, we print the dataset, to check that all our variables are correctly converted into floats before beginning to work with the data.

In [ ]:
# Check data types
print(DataProject_Data.dtypes)

# Inspect sample data
print(DataProject_Data.head())

# Calculate statistics
print(DataProject_Data.describe())

t           int64
DK_B_t    float64
DK_L_t    float64
DK_K_t    float64
DK_Q_t    float64
US_B_t    float64
US_L_t    float64
US_K_t    float64
US_Q_t    float64
IN_B_t    float64
IN_L_t    float64
IN_K_t    float64
IN_Q_t    float64
CN_B_t    float64
CN_L_t    float64
CN_K_t    float64
CN_Q_t    float64
dtype: object
      t       DK_B_t       DK_L_t        DK_K_t        DK_Q_t       US_B_t  \
0  1960  17203.33000  2111660.000  2.666270e+11  7.579978e+10  18524.19000   
1  1961  17461.37995  2123738.695  2.834730e+11  7.854689e+10  18802.05285   
2  1962  17723.30065  2135886.481  3.008210e+11  8.133756e+10  19084.08364   
3  1963  17989.15016  2148103.751  3.186760e+11  8.417337e+10  19370.34490   
4  1964  18258.98741  2160390.905  3.370440e+11  8.705587e+10  19660.90007   

        US_L_t        US_K_t        US_Q_t       IN_B_t       IN_L_t  \
0  70956940.00  1.469310e+13  3.510940e+12  1105.530000  179374600.0   
1  71921954.38  1.518190e+13  3.618440e+12  1122.112950  182836529.

# Exploring the data

**Descriptive statistics:**

In the cell below, we display the descriptive statistics of the dataset. As shown, all variable names start with the country code, followed by the name of the variable. E.g. DK_B_t states the technological stage in Denmark at given time t and vice versa. The descriptive statistics provide a summary of the distribution of various economic indicators across four countries (Denmark, the United States, India, and China) over a period from 1960 to 2100. 

In [ ]:
# Generate descriptive statistics for your DataFrame
descriptive_stats = DataProject_Data.describe()

# Display the descriptive statistics as a styled HTML table
# Set styling options for the table
descriptive_stats_table = descriptive_stats.style.set_table_styles([
    {'selector': 'th', 'props': [('background-color', 'white')]},  # Set background color for table headers
    {'selector': 'td', 'props': [('font-size', '12pt')]},         # Set font size for table data
    {'selector': 'tr:nth-of-type(odd)', 'props': [('background-color', 'whitesmoke')]},  # Alternate row color
    {'selector': 'tr:nth-of-type(even)', 'props': [('background-color', 'white')]}        # Alternate row color
]).format("{:.2f}")  # Format numerical values to have two decimal places

# Render the styled table
descriptive_stats_table

**Economic Disparities:**

The descriptive statistics highlight substantial differences in economic indicators such as the technological stage, GDP, capital stock, and population size among the four countries. In the table, it becomes clear, that we are working with both high- and low income countries. Even though the sizes in GDP are larger in India and China than in Denmark, GDP per capita is substantially lower due to the large population sizes in the two developing countries. Looking at the technological stages, it is clear that these are larger in the advanced countries than in India and China. This could impy that a change in the technological stage may have very different implications for advanced and developing countries. More on this later.

However, keeping all this in mind, remember that the dataset is based on a theoretical model and that all data (except initial values in 1960) is simulated and not actual values. Although, the simulations still provide some realistic insights into the real evolution of the variables, since parameter values for the simulations have been collected and calculated on country-specific levels. This means that, ceteris paribus, the evolution of the variables are close to the real and expected evolution in the variables. 

**Plotting and displaying the data:**

Below, we plot and display the variables in a figure, where we can choose what variable to look at and select wanted the time period. Beware of the label on the y-axis, where the variables $B_t$ is displayed in million US dollars, $L_t$ in millions, $Q_t$ in billion US dollars and $K_t$ in trillion US dollars.

In [ ]:
# Define colors for each country
colors = {'DK': 'blue', 'US': 'green', 'IN': 'red', 'CN': 'purple'}

# Define the function to plot the data for the selected indicator and year range
def plot_indicator(indicator, year_range):
    # Set up the plot figure with specified size
    plt.figure(figsize=(10, 6))
    
    # Determine scaling factor and y-axis label based on indicator
    if indicator == 'B_t' :
        scale_factor = 1_000  # Display in thousands
        ylabel = 'Value (in thousands)'
    elif indicator == 'L_t' :
        scale_factor = 1_000_000  # Display in millions
        ylabel = 'Value (in millions)'
    elif indicator == 'K_t':
        scale_factor = 1_000_000_000_000  # Display in trillions
        ylabel = 'Value (in Trillions)'
    elif indicator == 'Q_t':
        scale_factor = 1_000_000_000  # Display in billions
        ylabel = 'Value (in Billions)'
    else:
        scale_factor = 1  # No scaling
        ylabel = 'Value'
    
    # Iterate over each country and plot its data
    for country in ['DK', 'US', 'IN', 'CN']:
        # Retrieve the data for the current indicator
        data = DataProject_Data[country + '_' + indicator]
        
        # Convert values to desired scale
        data_scaled = data / scale_factor
        
        # Plot data for the current country with specified color
        plt.plot(DataProject_Data['t'], data_scaled, label=country, color=colors[country])
    
    # Set plot title, x-axis label, y-axis label, and legend
    plt.title('Trend of {} Indicator for Each Country Over Time'.format(indicator))
    plt.xlabel('Year')
    plt.ylabel(ylabel)
    plt.legend()
    
    # Add gridlines to the plot
    plt.grid(True)
    
    # Set x-axis limits based on the selected year range
    plt.xlim(year_range)
    
    # Display the plot
    plt.show()

# Get the list of indicators from the column names of the DataFrame
indicators = ['B_t', 'L_t', 'K_t', 'Q_t']

# Define the year range based on the minimum and maximum years in the data
min_year = DataProject_Data['t'].min()
max_year = DataProject_Data['t'].max()

# Create an interactive dropdown menu to select the indicator and a slider to select the year range
interact(plot_indicator, 
         indicator=indicators, 
         year_range=widgets.IntRangeSlider(min=min_year, max=max_year, step=1, 
                                           value=[min_year, max_year], description='Year Range'))

interactive(children=(Dropdown(description='indicator', options=('B_t', 'L_t', 'K_t', 'Q_t'), value='B_t'), In…

<function __main__.plot_indicator(indicator, year_range)>

The figure above displays the simulated trend of four indicators for each country, Denmark, The US, China and India. The four indicators are the technological stage, $B_t$, population sizes, $L_t$, physical capital, $K_t$ and GDP/output, $Q_t$. 

It is clear that the technological level is considerably larger for the countries, Denmark and the US, while the population level is larger for India and China. Further we observe that, according to the theoretical model, the population of India should surpass China in 2040. In reality, the population of India has already surpassed the population level of China in 2022. Thus, the model captures this, but underestimates it. The stock of physical capital is expected to increase for all countries, where Denmark (as the smallest country, geographically) has the smallest level of physical capital stock. As perhaps expected, China holds the largest stock of physical capital. Lastly, the US has the largest GDP, followed by China. Denmark has the lowest. However, remember that these figures are not given in per capita sizes, thus they don't tell us anything about the welfare of the individual. 

In [ ]:
# Define the list of countries
countries = ['DK', 'US', 'IN', 'CN']

# Define the dropdown widget for selecting the country
country_dropdown = widgets.Dropdown(
    options=countries,
    value='DK',
    description='Country:'
)

# Define the function to update the plot based on the selected country
def update_plot(country):
    # Calculate alpha for the selected country and update the plot
    alpha_col_name = f'{country}_alpha'
    alpha_value = alpha_calculator.calculate_alpha(f'{country}_Q_t', f'{country}_B_t', f'{country}_K_t', f'{country}_L_t')
    DataProject_Data[alpha_col_name] = alpha_value
    
    # Plot alpha values over time for the specified country
    plt.figure(figsize=(10, 6))  # Create a new figure for better organization
    plt.plot(DataProject_Data['t'], DataProject_Data[alpha_col_name], label=f'{country} Alpha', color='blue')
    plt.xlabel('Year')
    plt.ylabel('Alpha')
    plt.title('Country Specific Alpha over Time')
    plt.legend()
    plt.grid(True)
    plt.show()

# Create the interactive plot
interact(update_plot, country=country_dropdown)

interactive(children=(Dropdown(description='Country:', options=('DK', 'US', 'IN', 'CN'), value='DK'), Output()…

<function __main__.update_plot(country)>

In the figure above, the respective $\alpha$-values for each country are displayed. It is essential to look into the evolution of the $\alpha$-values since they are tightly correlated to the technological stage for each country. $\alpha$ determines how sensitive output is to changes in technology, since the parameter determines the elasticity of output with respect to $B_t$, $K_t$ and $L_t$. 

For Denmark, $\alpha_{DK}$ is increasing over the entire period with decreasing marginal returns, and the parameter converges towards $\alpha_{DK} = 0.512$.

For the United States, $\alpha_{US}$ is initially increasing up untill the mid 90's before decreasing untill year 2100, where $\alpha_{US} = 0.505$. 

For India, $\alpha_{IN}$ is also initially increasing up untill around 2020, from where the parameter value is decreasing. The end value for the parameter in 2100 is $\alpha_{IN} = 0.500$.

For China, $\alpha_{CN}$ is increasing up untill around 2040, from where the parameter value is slightly decreasing. The end value for the parameter in 2100 is $\alpha_{CN} = 0.505$.

This means that by year 2100 the gross domestic product $(Q_t)$ in Denmark is more sensitive to shocks in the technological stage than the other countries. High-income countries tends to invest more in research, education, and infrastructure, fostering innovation and technological advancement, which causes them to be more sensitive to changes in technologies.

# Technology increase in 2022 with 20%

In this section we are creating a new varibale called $B^{new}_t$, which inceases with 20% from 2022. After making the variable, we plot it in an interactive graph so we can see both $B_t$ and $B^{new}_t$ for each country. 

In [ ]:
# Define a condition for assigning values to B_new_t which makes it possible to only look at the period from 2022-2100
condition = (DataProject_Data['t'] >= 2022) & (DataProject_Data['t'] <= 2100)

# Create a dictionary to map country codes to their corresponding 'B_t' columns
country_b_columns = {
    'DK': 'DK_B_t',
    'US': 'US_B_t',
    'IN': 'IN_B_t',
    'CN': 'CN_B_t'
}

# Iterate over the countries and assign values to B_new_t based on the condition
for country, b_column in country_b_columns.items():
    # Assign values to B_new_t for the current country based on the condition
    DataProject_Data.loc[~condition, f'{country}_B_new_t'] = DataProject_Data[b_column]
    DataProject_Data.loc[condition, f'{country}_B_new_t'] = DataProject_Data[b_column] * 1.2


# Define countries and their corresponding B_t and B_new_t columns
countries = ['DK', 'US', 'IN', 'CN']
country_b_columns = {country: f'{country}_B_t' for country in countries}
country_b_new_columns = {country: f'{country}_B_new_t' for country in countries}

In [ ]:
# Define a function to plot B_t and B_new_t for a selected country
def plot_country_b_values(country):
    plt.figure(figsize=(10, 6))
    # Divide B_t and B_new_t values by 1,000 to display in thousands
    plt.plot(DataProject_Data['t'], DataProject_Data[f'{country}_B_t'] / 1_000, label=f'{country} B_t (thousands)')
    plt.plot(DataProject_Data['t'], DataProject_Data[f'{country}_B_new_t'] / 1_000, label=f'{country} B_new_t (thousands)')
    plt.title(f'{country} B_t vs B_new_t')
    plt.xlabel('Year')
    plt.ylabel('Value (thousands)')
    plt.legend()
    plt.grid(True)
    plt.show()

# Define countries and their corresponding B_t and B_new_t columns
countries = ['DK', 'US', 'IN', 'CN']

# Create dropdown menu for selecting a country
country_dropdown = widgets.Dropdown(options=countries, description='Country')

# Interact function to update plot based on dropdown selection
@interact(country=country_dropdown)
def update_plot(country):
    plot_country_b_values(country)

interactive(children=(Dropdown(description='Country', options=('DK', 'US', 'IN', 'CN'), value='DK'), Output())…

When we change $B_t$ to $B^{new}_t$, $Q_t$ (GDP) will change, all else equal. Therefore, we introduce $Q^{new}_t$, which mirrors $Q_t$ until 2021. From 2022 onwards, $Q^{new}_t$ is recalculated using all parameters in our Cobb-Douglas production function along with the new technology variable $B^{new}_t$. We display $Q_t$ and $Q^{new}_t$ in an interactive graph to visualize these changes further below. 

In [ ]:
# Create an instance of AlphaCalculator with the provided data
alpha_calculator = AlphaCalculator(DataProject_Data)

# Define the list of countries
countries = ['DK', 'US', 'IN', 'CN']

# Calculate Q_new_t and Q_t for each country using the alpha values
for country in countries:
    # Calculate alpha values for the current country
    alpha_col_name = f'{country}_alpha'
    alpha_value = alpha_calculator.calculate_alpha(f'{country}_Q_t', f'{country}_B_t', f'{country}_K_t', f'{country}_L_t')
    DataProject_Data[alpha_col_name] = alpha_value
    
    # Calculate Q_new_t using the provided formula
    B_new_col = f'{country}_B_new_t'
    K_col = f'{country}_K_t'
    L_col = f'{country}_L_t'
    DataProject_Data[f'{country}_Q_new_t'] = (DataProject_Data[B_new_col] ** alpha_value) * ((DataProject_Data[K_col] * DataProject_Data[L_col]) ** (1 - alpha_value))
    
    # Calculate Q_t using the same formula with B_t instead of B_new_t
    B_col = f'{country}_B_t'
    DataProject_Data[f'{country}_Q_t'] = (DataProject_Data[B_col] ** alpha_value) * ((DataProject_Data[K_col] * DataProject_Data[L_col]) ** (1 - alpha_value))

In [ ]:
# Define a function to plot Q_t and Q_new_t for a selected country
def plot_country_q_values(country):
    plt.figure(figsize=(10, 6))
    # Convert Q_t and Q_new_t values to billions
    plt.plot(DataProject_Data['t'], DataProject_Data[f'{country}_Q_t'] / 1e9, label=f'{country} Q_t (billions)', color='blue')
    plt.plot(DataProject_Data['t'], DataProject_Data[f'{country}_Q_new_t'] / 1e9, label=f'{country} Q_new_t (billions)', color='orange')
    plt.title(f'{country} Q_t vs Q_new_t')
    plt.xlabel('Year')
    plt.ylabel('Value (billions)')
    plt.legend()
    plt.grid(True)
    plt.show()

# Define countries and their corresponding Q_t and Q_new_t columns
countries = ['DK', 'US', 'IN', 'CN']

# Create dropdown menu for selecting a country
country_dropdown = widgets.Dropdown(options=countries, description='Country')

# Interact function to update plot based on dropdown selection
@interact(country=country_dropdown)
def update_plot(country):
    plot_country_q_values(country)

interactive(children=(Dropdown(description='Country', options=('DK', 'US', 'IN', 'CN'), value='DK'), Output())…

In the diagram above, it becomes clear how an increase in the technology stage has a positive effect on output. Even though we use the same measure for each country (billion USD), it is quite hard to determine for which country the largest absolute effect occurs. In order to determine this, we can calculate the absolute difference between the old output, $Q_t$ and the new output $Q^{new}_t$ and then calculate how large the percentage change in output is in average:

In [ ]:
def calculate_average_annual_difference(country):
    try:
        Q_diff = DataProject_Data[f'{country}_Q_new_t'] - DataProject_Data[f'{country}_Q_t']
        avg_annual_diff = Q_diff.mean()
        return avg_annual_diff
    except KeyError as e:
        print(f"Error: {e}")
        return None

def calculate_average_annual_percentage_increase(country):
    try:
        Q_t = DataProject_Data[f'{country}_Q_t']
        Q_new_t = DataProject_Data[f'{country}_Q_new_t']
        Q_diff = Q_new_t - Q_t
        percentage_increase = (Q_diff / Q_t) * 100
        avg_annual_percentage_increase = percentage_increase.mean()
        return avg_annual_percentage_increase
    except KeyError as e:
        print(f"Error: {e}")
        return None

# Define countries
countries = ['DK', 'US', 'IN', 'CN']

# Calculate and print the average annual absolute increase for each country
for country in countries:
    avg_annual_diff = calculate_average_annual_difference(country)
    if avg_annual_diff is not None:
        print(f"The average annual absolute increase from Q_t to Q_new_t for {country} is {avg_annual_diff:.2f}")

# Calculate and print the average annual percentage increase for each country
for country in countries:
    avg_annual_percentage_increase = calculate_average_annual_percentage_increase(country)
    if avg_annual_percentage_increase is not None:
        print(f"The average annual percentage increase from Q_t to Q_new_t for {country} is {avg_annual_percentage_increase:.2f}%")


The average annual absolute increase from Q_t to Q_new_t for DK is 56751086748.34
The average annual absolute increase from Q_t to Q_new_t for US is 4355866353657.45
The average annual absolute increase from Q_t to Q_new_t for IN is 1914990088053.13
The average annual absolute increase from Q_t to Q_new_t for CN is 3495480655900.84
The average annual percentage increase from Q_t to Q_new_t for DK is 5.48%
The average annual percentage increase from Q_t to Q_new_t for US is 5.41%
The average annual percentage increase from Q_t to Q_new_t for IN is 5.36%
The average annual percentage increase from Q_t to Q_new_t for CN is 5.41%


The results shows that the technological increase has the largest effect on the average annual absolute change in output in the US, followed by China, India and lastly Denmark. However, when looking into the average annual percentage change, the effect is largest in Denmark, followed by the US and China, and lastly India. 

# Impact of the technological increase

**Per worker terms**:

To analyse the impact of the technological increase, it becomes essential to look at the impact for the individual worker. In this case, it means we should look into the effects of the technological increase on output per worker. We do so by introducing two new variables, $q_t = \frac{Q_t}{L_t}$ and $q^{new}_t = \frac{Q^{new}_t}{L_t}$. First, we calculate the two variables and below they are presented in a diagram for each country over the entire period. 

In [ ]:
# Define the list of countries
countries = ['DK', 'US', 'IN', 'CN']

# Calculate q_t = Q_t / L_t for each country and assign it to a new column
for country in countries:
    DataProject_Data[f'{country}_q_t'] = DataProject_Data[f'{country}_Q_t'] / DataProject_Data[f'{country}_L_t']
    DataProject_Data[f'{country}_q_new_t'] = DataProject_Data[f'{country}_Q_new_t'] / DataProject_Data[f'{country}_L_t']


def plot_q_t(year_range, country):
    start_year, end_year = year_range
    plt.figure(figsize=(10, 6))
    mask = (DataProject_Data['t'] >= start_year) & (DataProject_Data['t'] <= end_year)
    plt.plot(DataProject_Data['t'][mask], DataProject_Data[f'{country}_q_t'][mask], label=f'{country} q_t')
    plt.plot(DataProject_Data['t'][mask], DataProject_Data[f'{country}_q_new_t'][mask], label=f'{country} q_new_t')
    plt.xlabel('Year')
    plt.ylabel('Values')
    plt.title(f'q_t and q_new_t Values for {country} from {start_year} to {end_year}')
    plt.legend()
    plt.grid(True)
    plt.show()

year_range_slider = widgets.IntRangeSlider(
    value=[1960, 2100],  # Adjusted to fit the example data range
    min=1960,
    max=2100,
    step=1,
    description='Year Range:',
    continuous_update=False
)

# Create dropdown menu for selecting a country
country_dropdown = widgets.Dropdown(options=countries, description='Country')

# Interact function to update plot based on dropdown selection
@interact(year_range=year_range_slider, country=country_dropdown)
def update_plot(year_range, country):
    plot_q_t(year_range, country)

interactive(children=(IntRangeSlider(value=(1960, 2100), continuous_update=False, description='Year Range:', m…

In [ ]:
def calculate_average_annual_difference(country):
    try:
        q_diff = DataProject_Data[f'{country}_q_new_t'] - DataProject_Data[f'{country}_q_t']
        avg_annual_diff = q_diff.mean()
        return avg_annual_diff
    except KeyError as e:
        print(f"Error: {e}")
        return None

def calculate_average_annual_percentage_increase(country):
    try:
        q_t = DataProject_Data[f'{country}_q_t']
        q_new_t = DataProject_Data[f'{country}_q_new_t']
        q_diff = q_new_t - q_t
        percentage_increase = (q_diff / q_t) * 100
        avg_annual_percentage_increase = percentage_increase.mean()
        return avg_annual_percentage_increase
    except KeyError as e:
        print(f"Error: {e}")
        return None

# Define countries
countries = ['DK', 'US', 'IN', 'CN']

# Calculate and print the average annual absolute increase for each country
for country in countries:
    avg_annual_diff = calculate_average_annual_difference(country)
    if avg_annual_diff is not None:
        print(f"The average annual absolute increase from q_t to q_new_t for {country} is {avg_annual_diff:.2f}")

# Calculate and print the average annual percentage increase for each country
for country in countries:
    avg_annual_percentage_increase = calculate_average_annual_percentage_increase(country)
    if avg_annual_percentage_increase is not None:
        print(f"The average annual percentage increase from q_t to q_new_t for {country} is {avg_annual_percentage_increase:.2f}%")

The average annual absolute increase from q_t to q_new_t for DK is 14325.76
The average annual absolute increase from q_t to q_new_t for US is 13566.91
The average annual absolute increase from q_t to q_new_t for IN is 1228.37
The average annual absolute increase from q_t to q_new_t for CN is 2850.39
The average annual percentage increase from q_t to q_new_t for DK is 5.48%
The average annual percentage increase from q_t to q_new_t for US is 5.41%
The average annual percentage increase from q_t to q_new_t for IN is 5.36%
The average annual percentage increase from q_t to q_new_t for CN is 5.41%


Since India and China has such huge populations, the results changes when looking into per worker terms. Remember from earlier, that we found that the average annual absolute change in output were largest in the US, followed by China, India and Denmark. When looking at per worker terms, the largest effect is observed for Denmark, followed by the US, China and India. In percentage terms, the same results still hold from earlier. These results show us, that the technological increase has the largest overall effect in level in the US, but when looking at per worker terms and individual welfare, the effect is largest in Denmark. 

# Conclusion

In this paper, we found that the long run effects of a positive shock to technology of $20\%$ have positive effects on output per worker for all four countries. The results are thus consistent across high- and low-income countries, however, the impacts are different. 

We found that the growth rates for output per worker are not sentitive to shocks, but the levels of output per worker are. This means, that the absolut values changes, but not the speed they increase with. We found that when looking only at the gap in total output, the largest effect were observed for the US, followed by China, India and Denmark. However, when looking at the gaps between the curves for output per worker with and without a shock to technology for each country, that the largest gaps are seen in the two high-income countries. For Denmark, the gap is $14,325.76$ USD, which is the largest gap. This is followed by $13,566.91$ USD for the United States, $2,850.39$ USD for China and $1,228.37$ USD for India. 

This means that the long run effects of a positive schock to technology are positive on output per worker, but initial conditions such as population size and output size has great effects on the results. 

# Bibliography on data

### Capital Stock

University of Groningen and University of California, Davis "Capital Stock at Constant National Prices".  
Retrieved from FRED, Federal Reserve Bank of St. Louis

**Denmark**: [RKNANPDKA666NRUG](https://fred.stlouisfed.org/series/RKNANPDKA666NRUG)  
**The United States**: [RKNANPUSA666NRUG](https://fred.stlouisfed.org/series/RKNANPUSA666NRUG)  
**India**: [RKNANPINA666NRUG](https://fred.stlouisfed.org/series/RKNANPINA666NRUG)  
**China**: [RKNANPCNA666NRUG](https://fred.stlouisfed.org/series/RKNANPCNA666NRUG)

### Gross Domestic Product

University of Groningen and University of California, Davis "Real GDP at Constant National Prices".  
Retrieved from FRED, Federal Reserve Bank of St. Louis

**Denmark**: [RGDPNADKA666NRUG](https://fred.stlouisfed.org/series/RGDPNADKA666NRUG)  
**The United States**: [RGDPNAUSA666NRUG](https://fred.stlouisfed.org/series/RGDPNAUSA666NRUG)  
**India**: [RGDPNAINA666NRUG](https://fred.stlouisfed.org/series/RGDPNAINA666NRUG)  
**China**: [RGDPNACNA666NRUG](https://fred.stlouisfed.org/series/RGDPNACNA666NRUG)

### Labour Force

University of Groningen and University of California, Davis "Number of Persons Engaged".  
Retrieved from FRED, Federal Reserve Bank of St. Louis

**Denmark**: [EMPENGDKA148NRUG](https://fred.stlouisfed.org/series/EMPENGDKA148NRUG)  
**The United States**: [EMPENGUSA148NRUG](https://fred.stlouisfed.org/series/EMPENGUSA148NRUG)  
**India**: [EMPENGINA148NRUG](https://fred.stlouisfed.org/series/EMPENGINA148NRUG)  
**China**: [EMPENGCNA148NRUG](https://fred.stlouisfed.org/series/EMPENGCNA148NRUG)

## Technological Stage
This variable has been calculated using the other available variables for each country for each year. 

## Other parameters and variables
We want to mention once again, that all data except the initial year 1960 has been simulated through a Green Solow Model used in a previous bachelor thesis, where many other parameters has been collected or calculated. Among these are savings rates, depreciation rates on capital etc. However, the most important variables and the data retrieved for these are listed above. 
